### ISSUE TITLE
Huge latency when seeking/playing

### ISSUE THREAD

**Issue by bebehei** [librespot-org/librespot#158]
Hi guys,

I've built librespot successfully with pulseaudio support. It also works reasonably well.

I had been using Spotify Connect already for a long time with my Smartphone and computer together and in comparison to this, there is a small thing which bugs me.

Seeking and switching tracks makes the audio output stop for multiple seconds before the track starts or jumps to the seeked point.

The log reads the following:

```
[...]
INFO:librespot_playback::player: Track "Get Me To The Church On Time - Live At The Sands Hotel And Casino/1966" loaded
ERROR:librespot_playback::player: Could not write audio: Invalid argument
[...]
DEBUG:librespot_audio::fetch: File 5ce8e3b0c765502e75c3172090f1bb5e17e38fc9 complete, saving to cache
DEBUG:librespot_connect::spirc: kMessageTypePrev "butter" 4428af77be5ecd3a401356eec0e155b661993dec 8 0
DEBUG:librespot_playback::player: command=Seek(0)
ERROR:librespot_playback::player: Could not write audio: Invalid argument
[...]
```

There is no difference, if I change the bitrate/caching options.

Also interestingly, there is no delay, if I pause and play the track again.

Reproducible on ArchLinux on my desktop (amd64) and on an Odroid C2 (aarch64).

**Comment by bebehei** [comment]
Ok, I've got a workaround now. I'm using the `pipe` backend now and I'm piping everything into `pacat`. Now librespot works like a usual Spotify Connect and doesn't have these huge latencies, but I'd still favor a native working pulseaudio support.

**Comment by lassos** [comment]
Have the same problem. Using pulseaudio. Huge delay with seeking and switching forward or back

**Comment by lassos** [comment]
I have no kodi. Just debian with pulseaudio. And in default.pa is nothing with keeper audio sinken alive. In original librespot i have never had seek play latency

**Comment by bebehei** [comment]
@lassos I don't have kodi either. Could you please elaborate?

**Comment by lassos** [comment]
https://github.com/librespot-org/librespot/issues/180

So in this thread the talk about 

Try disabling option "keep audio sink alive" in Kodi


So where can i do it. In pulseaudio config ---system.pa

So what do you need.

**Comment by ComlOnline** [comment]
@lassos depending on how old your version of librespot was we implemented a change in the volume in #10 So that its no longer linear but logarithmic. 

@bebehei I'll build a recent version with pulse audio and have a look tomorrow.

**Comment by bebehei** [comment]
@ComlOnline I'm regularly rebuilding librespot to match current master. I've even started using librespot **after** #10 had been merged.

**Comment by MarcelKr** [comment]
@lassos @bebehei My issue wasn't related to Kodi. I was not a member of the ``audio`` groupd and therefore I had no sound. Now that I fixed that, I get the exact symptoms seen in this issue. E.g. playback working but with the ``Could not write audo`` error message and the delay when seeking.

**Comment by lassos** [comment]
Oh , OK. I am getting the same error -> Could not write audio 

`mpd@-XXX:/var/logRUST_BACKTRACE=1 /usr/bin/librespot --username XXX --password XXX --bitrate 320 --name XXX --backend pulseaudio
INFO:librespot: librespot 593dfa0 (2018-02-27). Built on 2018-02-28. Build ID: heiwVZ8W
INFO:librespot_core::session: Connecting to AP "gew1-accesspoint-b-pnns.ap.spotify.com:4070"
INFO:librespot_core::session: Authenticated as "XXX" !
INFO:librespot_core::session: Country: "DE"
INFO:librespot_playback::player: Loading track "Norrsken"
INFO:librespot_playback::player: Track "Norrsken" loaded
ERROR:librespot_playback::player: Could not write audio: Invalid argument
^[[A
ERROR:librespot_playback::player: Could not write audio: Invalid argument`

it would be great to get a solution cause it's really uncomftable to wait few seconds to seek or play next titles. thanks lars

**Comment by bebehei** [comment]
Like I said it: Use `librespot --backend pipe | pacat` as a workaround. This helps and lets the issue go away. It doesn't change the fact, that librespot's pulseaudio support is broken, but it works :tm: 

If you need a full writeup, [you can find it on my webpage](https://bebehei.de/odroid-mediacenter/#librespot).

**Comment by lassos** [comment]
pretty cool, i have overlooked it, i am sorry. worked  for me .-) 

thanks a lot bebehei

**Comment by lassos** [comment]
Since today i got another error : 

`ERROR:librespot_playback::player: Could not write audio: Broken pipe (os error 32)
ERROR:librespot_playback::player: Could not write audio: Broken pipe (os error 32)
`

librespot does not work anymore with pulse

**Comment by sashahilton00** [comment]
This is somewhat present on portaudio. What I believe the problem is, is that Track.load is called whenever a song is selected, even if only briefly when skipping through them. Hence, if you skip 5 songs, librespot loads 5 songs before the one you actually want. The solution is to implement an audio_queue, which will hopefully happen at some point. This leads to the loading latency. With regards to the seeking, that sounds like a CPU limitation, given that the pipe works.

**Comment by bebehei** [comment]
I found out, that @brain0 fixed this in meantime with PR #194. Update to current master to have the changes included.

ping @Obrekr, you may want to update your commits again.
